Social: Update SIG init logic and update plan checks to check if SIG is enabled

spsiddarthan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->


## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Moved SIG init to admin_init
* Updated the `is_social_image_generator_enabled` function in `class-publicize-base` to use `Current_Plan`
* Added some code in `class-publicize-setup` to refresh plan data on every `jetpack_heartbeat`

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pdrWKz-P7-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No
## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Make sure no blog stickers are added to the site
* Purchase the advanced plan on your Jetpack site and add the blog sticker for SIG. 
* On your localhost, run `jetpack docker wp shell` and inside the shell, execute `do_action('jetpack_heartbeat');`. 
* Go to your publicize sidebar on the editor and see that SIG is enabled. 
* Remove the blog sticker 
* On your localhost, run `jetpack docker wp shell` and inside the shell, execute `do_action('jetpack_heartbeat');`. 
* Go to your publicize sidebar on the editor and see that SIG is not enabled.

spsiddarthan committed:
[not verified] Social: Update SIG initilization. Update plan check to see if Sig is enabled.

spsiddarthan committed:
[not verified] changelog

spsiddarthan committed:
[not verified] Move plan refresh logic to the plans package.

spsiddarthan committed:
[not verified] changelog

spsiddarthan committed:
[not verified] Update lock files.

spsiddarthan committed:
[not verified] Changelogger validity.

pablinos commented on projects/packages/plans/src/class-current-plan.php:
```diff
@@ -211,7 +228,6 @@ private static function store_data_in_option( $option, $data ) {
 	 */
 	public static function refresh_from_wpcom() {
 		// Make the API request.
-
```
Nitpick! Do we want to add a newline here?

pablinos commented on projects/packages/plans/src/class-current-plan.php:
```diff
@@ -364,4 +380,11 @@ function_exists( 'wpcom_feature_exists' ) &&
+	/**
+	 * Refreshess the plan data.
+	 */
+	public static function refresh_plan_data() {
```
Shall we just hook into `refresh_from_wpcom` instead? We could rename that if `refresh_plan_data` is more appropriate.

pablinos commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -307,7 +307,7 @@ class_exists( 'Jetpack' ) ||
 					),
 					'hasPaidPlan'                   => $publicize->has_paid_plan(),
```
For some reason I didn't need to make the call to `jetpack_heartbeat` after removing the blog sticker, so the plan data got refreshed anyway. I thought maybe we were doing it here, when checking for the paid plan, but it doesn't look like it. We do [force a check here](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/social/src/class-jetpack-social.php#L226) but that should only run on the main admin page. That does have the side effect we're looking for though, of forcing a refresh when we load the admin page.

spsiddarthan replied:
I think the force check runs both in the editor and admin page, for the social plugin. 

Did you test using the jetpack plugin or the social plugin?

Edit: This happens due to a hook `do_action('current_screen'), WP_Hook->do_action, WP_Hook->apply_filters, Automattic\Jetpack\Publicize\Publicize_Setup::init_sharing_limits, Automattic\Jetpack\Publicize\Publicize_Base->has_paid_plan, Automattic\Jetpack\Current_Plan::supports, Automattic\Jetpack\Current_Plan::refresh_from_wpcom`, however, there is some caching setup in it. 

```
	/**
	 * Check if we have a paid Jetpack Social plan.
	 *
	 * @param bool $refresh_from_wpcom Whether to force refresh the plan check.
	 *
	 * @return bool True if we have a paid plan, false otherwise.
	 */
	public function has_paid_plan( $refresh_from_wpcom = false ) {
		static $has_paid_plan = null;
		if ( ! $has_paid_plan ) {
			$has_paid_plan = Current_Plan::supports( 'social-shares-1000', $refresh_from_wpcom );
		}
		return $has_paid_plan;
	}
	
```

pablinos replied:
After testing this again with just Jetpack enabled, then the plan information isn't being refreshed on editor load, which is as expected.

pablinos commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -307,7 +307,7 @@ class_exists( 'Jetpack' ) ||
 					),
 					'hasPaidPlan'                   => $publicize->has_paid_plan(),
 					'isEnhancedPublishingEnabled'   => $publicize->is_enhanced_publishing_enabled( Jetpack_Options::get_option( 'id' ) ),
```
Something for a follow-up I should think, but could we similarly sway this check to a `Current_Plan::supports` call?

spsiddarthan replied:
Yeah, sounds good! I have added it here - 1202583289493814-as-1204237958606623

pablinos reviewed (commented):
This is working well, but I don't need to call `jetpack_heartbeat` to get the state to change, which makes me think we're somehow forcing the plan data update on each editor load already. I can see an XMLRPC call happening, so it might be to do with that.

I've clicked around loading frontend pages etc, and I can't see any extra API calls as was happening before, so I think this does fix the problem.

Just a couple of minor comments on the code, and if we can get the e2e tests passing too, that would be great.

spsiddarthan committed:
[not verified] Code review comments.

spsiddarthan commented:
Even on jetpack, things are working fine for me, without me having to manually trigger the heartbeat. Wonder how!

spsiddarthan commented:
@pablinos, one end to end with boost is failing due to a timeout and I found out that when publicize gets setup, we have hook, which refreshes the plan data. It seems ok to me, but may be we can think of cleaning it up in a follow-up, if need be? 

Otherwise, I think I've addressed the comments on this PR. 

```
/**
	 * Check if we have a paid Jetpack Social plan.
	 *
	 * @param bool $refresh_from_wpcom Whether to force refresh the plan check.
	 *
	 * @return bool True if we have a paid plan, false otherwise.
	 */
	public function has_paid_plan( $refresh_from_wpcom = false ) {
		static $has_paid_plan = null;
		if ( ! $has_paid_plan ) {
			$has_paid_plan = Current_Plan::supports( 'social-shares-1000', $refresh_from_wpcom );
		}
		return $has_paid_plan;
	}
```

danielpost commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -44,6 +46,10 @@ public static function on_jetpack_feature_publicize_enabled() {
 
 		add_action( 'current_screen', array( static::class, 'init_sharing_limits' ) );
+
+		add_action( 'admin_init', array( static::class, 'init_social_image_generator' ) );
```
Hooking into the `admin_init` action breaks the `wp_after_insert_post` hook—it no longer runs for the actual post, only for the auto draft.

pablinos replied:
I've moved the plan check into the `wp_after_insert_post` hook, and added a check to make sure that the post is `publicizeable`. That's not a perfect check, but I think it prevents us setting the meta on any custom post types that might be registered on the site.

pablinos committed:
[not verified] Move the plan check into the wp_after_insert_post hook.

spsiddarthan committed:
[not verified] Make tests pass.

spsiddarthan committed:
[not verified] Added current plan as a dependency.

spsiddarthan committed:
[not verified] fix changelogger validity.

spsiddarthan committed:
[not verified] changelogger validity.

spsiddarthan committed:
[not verified] Added comment

spsiddarthan committed:
[not verified] Get rid of extra method.

spsiddarthan committed:
[not verified] Added default parma to the base class function.

spsiddarthan committed:
[not verified] Added default parma to the base class function.

spsiddarthan committed:
[not verified] Revert "Added default parma to the base class function."

This reverts commit ecde75e44ba4f8679942d068b64bcc73af48c9b4.

spsiddarthan committed:
[not verified] Rename method.

spsiddarthan committed:
[not verified] Rename method.

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1615,17 +1615,10 @@ public function is_enhanced_publishing_enabled( $blog_id ) {
-		}
-
-		return ! empty( $data['is_social_image_generator_enabled'] );
+	public function has_social_image_generator_feature() {
```
We've renamed this function as it gets overridden on WPCOM, and the signature changing was causing problems.

pablinos commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -45,6 +47,9 @@ public static function on_jetpack_feature_publicize_enabled() {
 
 		add_action( 'current_screen', array( static::class, 'init_sharing_limits' ) );
+
+		Current_Plan::init();
```
As we discussed, I think we might want to move this to Jetpack Social, so that we don't refresh the plan data twice on the `jetpack_heartbeat` when Jetpack has Publicize enabled.

pablinos reviewed (commented):
I think this is pretty much there now. If we move the `Current_Plan::init()` then it should be good to go.

spsiddarthan committed:
[not verified] Register the jetpack_heartbeat on the social plugin.

pablinos committed:
Remove Plans package change.

pablinos committed:
Update composer.lock

pablinos committed:
Update composer.lock
